### DISCUSSION TITLE
Issue with injecting planet, square artifact

### DISCUSSION THREAD

**Discussion by lwelzel** [vortex-exoplanet/VIP#590] [Q&A]
### Problem description:

I am trying to inject a synthetic planet (spectrum) using cube_inject_companions in fakecomp.py. Overall this works very well, however when using a high contrast and a specific instrument psf (spectrum) I get the following square artifact in the image:

![image](https://user-images.githubusercontent.com/29613344/236633352-171c5e61-98ae-4124-9c74-ed9b4d7b3157.png)

The square persists with rotation and is dependent on the wavelength, see below:

![image](https://user-images.githubusercontent.com/29613344/236633523-a38a6163-9fa9-48ad-8cdf-a88345282b53.png)

I use the normalize_psf function as recommended in the docstring of cube_inject_companions. While I am only visually seeing it with high contrast ~1e-3, I assume that this persistent no matter the contrast.  Since this is the same size as my psf I assume that this is due to the way the image cubes of star and companion are added.

Has anybody come up against something similar and found a solution to the sharp edge? I expect this to become a problem when training a (C)NN on synthetic images produced with cube_inject_companions since the co-moving sharp edge will be a possible feature to detect.

#### Median edge subtraction
One potential issue is that the psf image is "small" compared to its extent so that the outer pixels are far from zero (wavelength dependent):
![image](https://user-images.githubusercontent.com/29613344/236634620-a6239bb6-6fb0-4404-ac82-e22add428aba.png)

Subtracting the median of the edges of the image cube (see below) has major bad vibes since I am just shifting all the pixel values down so I am slightly changing the PSF. It helps it bit with the edge issue though. 

``` python
def fix_psf_median_corner(psf, **kwargs):
    median_edge = np.median([
        psf[:, 0, :],
        psf[:, -1, :],
        psf[:, :, 0],
        psf[:, :, -1],
    ])
    return psf - median_edge
```

![image](https://user-images.githubusercontent.com/29613344/236634836-47a51590-8595-402d-8a34-9c30db834f2d.png)

#### Padded psf image
Padding the psf first helps slightly but naturally preserves part of the outline square as the psf is not extended into the padding.

``` python
def fix_psf_pad(psf, padding=10, overlap=2):
    assert padding > overlap

    nz, ny, nx = psf.shape
    cy, cx = frame_center(psf)

    big_psf = np.zeros(shape=(nz, ny+2*padding, nx+2*padding))
    big_psf[:, padding:-padding, padding:-padding] = psf

    big_psf_mask = np.zeros(shape=(nz, ny+2*padding, nx+2*padding))
    big_psf_mask[:, padding-overlap:-padding+overlap, padding-overlap:-padding+overlap] = 1

    return big_psf
```

![image](https://user-images.githubusercontent.com/29613344/236639090-af1654dc-ee32-4a39-8001-86ee699b1525.png)

Does anyone have a solution/workaround for this problem for which I can still use the same psf? Any suggestions are very welcome.

**Comment by digirak** [comment]
Hi iwelzel,

Thanks for this issue, indeed we have seen this and its most prominently present for low contrasts.

I would recommend shaping your PSF so that only the central region where the companion is inserted (i.e only the pixels corresponding to the FWHM) are inserted and the rest is zero.

FWIW, it does not hurt training a deep learning algorithm much but try to avoid low contrasts and work in the high contrast regime.

Hope that helps

**Reply by digirak** [reply]
I also see your PSF is not circular so maybe you can try convolving a Gaussian kernel on your data to get a circular PSF?

**Reply by lwelzel** [reply]
Hi digirak,

thank you for the quick reply and suggestion!

**Comment by VChristiaens** [comment]
Hi @lwelzel, for the injection to work well visually the average pixel values at the edges of the PSF inset needs to be ~0 and to have a variance after scaling that is significantly smaller than local stellar speckle intensities.
A first important check is to make sure the average background level is indeed ~0 - to be checked in psf images before cropping. Any departure from 0 will have an increasing effect for injections at shallower contrasts. Note that if you're dealing with data from the exoplanet imaging data challenge, this was double-checked by the organizing team ;-)
If the background is correctly subtracted, the next thing to test is to cut wider images, to be sure to leave some space for background beyond the PSF wings (e.g. PSFs achieving a low Strehl ratio are more spread spatially, and would require a larger cut to capture more of the wing signals). 
If the injected PSF images still do not fit well in the injected cube images, it may mean the background noise becomes commensurate with the local stellar halo level after scaling the image to the desired injection level (which by eye may be the case for the examples you show, although I'd first try larger cuts). In that case, the best may indeed be to follow @digirak's suggestions, and either mask the PSF beyond a certain radius (if round), or threshold the PSF images - such that values beyond the PSF core are set to 0.

**Reply by lwelzel** [reply]
Hello @VChristiaens and thank you for the in-depth response. 
Do I understand correctly that with "cutting wider images" you mean cropping a larger psf less (e.g. 128x128 -> 64x64 instead of cropping to 32x32)? Is there a specific measure wrt to uniformity at the edges of the psf or relative variance compared to the intended injection location that is good practice? 
The images I attached are indeed from EIDC data, specifically the tutorial psf (psf_cube_sphere0.fits).

**Reply by VChristiaens** [reply]
Yes exactly: cropping less the original PSF image to not crop PSF wings by leaving a bit of background.

It's probably fair to assume that having between 1 and 2 orders of magnitude lower variance in the (scaled) PSF background noise than in the local stellar speckle noise at the intended injection location is what you'd want to avoid a noticeable transition at the edge of the injected PSF image. That being said, there may be another way to measure uniformity - I haven't dived specifically into that.

Great that you're considering participating to the EIDC by the way!
